src/test/util/mining.cpp

```diff
@@ -82,7 +82,8 @@ std::shared_ptr<CBlock> PrepareBlock(const NodeContext& node, const CScript& coi
             ->block);
 
     LOCK(cs_main);
-    block->nTime = Assert(node.chainman)->ActiveChain().Tip()->GetMedianTimePast() + 1;
+    const CBlockIndex* tip = Assert(node.chainman)->ActiveChain().Tip();
+    block->nTime = tip->nHeight == 0 ? GetTime<std::chrono::seconds>().count() : tip->GetBlockTime() + 1;
```

Comment:
> Is there a reason for the first commit?

It is because this is the only place where we set the chain timestamps to be in the past. In the form of
```
block_1_timestamp = genesis + 1
block_2_timestamp = median(genesis, block_1) + 1
block_3_timestamp = median(genesis, block_1, block_2) + 1
```
And in the wallet, we use the clock time to set the descriptors/keys creation time.

So, even when we create a wallet prior mining the chain, the wallet will always have a birth time after it. Which will lead to the wallet skipping those blocks during sync. This happens on the `wallet_balance.cpp` benchmark.

But.. I just went deeper over it and.. surprise surprise, seems that the functions that create blocks in the past: `generatetoaddress` (which is not the RPC command) and `PrepareBlock` are only used in benchmarks: `wallet_balance.cpp` and `block_assemble.cpp`.. so we can actually re-work this stuff differently cleaning some stuff.

The rest of the code uses `BlockAssembler::CreateNewBlock`, which sets the block time to the current, or mocked, clock time. So, no functional, unit or fuzz tests should be affected by this change.

Reply:
Ended up pushing a simpler approach. Focused on correcting `wallet_balance.cpp` only.

In the future, we could generate a fake chain instead of creating a "real" one. Similar to how it's done
in the `wallet_create_tx.cpp` benchmark.

Reply:
ha, loved it. I was so deep that ended up complicating things for no reason.
Dropped the first commit and squashed the mock time change into the last one.